lib/elixir/lib/module.ex

```diff
@@ -1221,12 +1221,29 @@ defmodule Module do
     behaviours = :ets.lookup_element(table, :behaviour, 2)
     impls = :ets.lookup_element(table, {:elixir, :impls}, 2)
 
+    check_behaviours(env, behaviours, all_definitions)
+
     if impls != [] do
       non_implemented_callbacks = check_impls(behaviours, impls)
       warn_missing_impls(env, non_implemented_callbacks, all_definitions, overridable_pairs)
     end
   end
 
+  defp check_behaviours(env, behaviours, all_definitions) do
+    Enum.each(behaviours, fn behaviour ->
+      cond do
+        function_exported?(behaviour, :behaviour_info, 1) ->
+          Enum.each(behaviour.behaviour_info(:callbacks), fn callback ->
+            nil
+          end)
+        function_exported?(behaviour, :module_info, 0) ->
+          :elixir_errors.warn(env.line, env.file, "@behaviour #{inspect behaviour} does not contain callbacks (for module #{inspect env.module})")
+        true ->
```

Comment:
Instead of having it at the end, let's have this as the first clause:

```elixir
not Code.ensure_compiled?(behaviour) ->
```

This way we can completely remove this clause here: https://github.com/brodeuralexis/elixir/blob/ebe61cfcb893841253ca607f88062246af35257b/lib/elixir/lib/module.ex#L1458